src/rtc_pl031.rs

```diff
@@ -0,0 +1,639 @@
+    // every day or on a certain day. It can sometimes be found in
+    // the 'Power Management' section of a motherboard's BIOS setup.
+    // This is not currently implemented, so we raise an error.
+    // TODO: Implement the match register functionality.
+    mr: u32,
+
+    // The interrupt mask.
+    imsc: u32,
+
+    // The raw interrupt value.
+    ris: u32,
+
+    // Used for notifying the driver about events.
+    interrupt_evt: T,
+}
+
+/// Errors encountered while handling RTC operations.
+#[derive(Debug)]
+pub enum Error<E> {
+    /// Failed to trigger interrupt.
+    Trigger(E),
+    /// Attempting to write to a non-existent register offset
+    /// or to the match register, which is not emulated.
+    BadWriteOffset(u64),
+}
+
+impl<T: Trigger> RTC<T> {
+    /// Creates a new `AMBA PL031 RTC` instance which uses `trigger` object
+    /// to notify the driver about new events.
+    ///
+    /// # Arguments
+    /// * `trigger` - The Trigger object that will be used to notify the driver
+    ///               about events.
+    ///
+    /// # Example
+    ///
+    /// You can see an example of how to use this function in the
+    /// [`Example` section from `RTC`](struct.RTC.html#example).
+    pub fn new(trigger: T) -> RTC<T> {
+        RTC {
+            // Counts up from 1 on reset at 1Hz (emulated).
+            counter: Instant::now(),
+
+            // The load register is initialized to zero.
+            lr: 0,
+
+            // The match register is initialised to zero (not currently used).
+            // TODO: Implement the match register functionality.
+            mr: 0,
+
+            // The interrupt mask is initialised as not set.
+            imsc: 0,
+
+            // The raw interrupt is initialised as not asserted.
+            ris: 0,
+
+            interrupt_evt: trigger,
+        }
+    }
+
+    /// Provides a reference to the interrupt event object.
+    pub fn interrupt_evt(&self) -> &T {
+        &self.interrupt_evt
+    }
+
+    // TODO: Reimplement with the match register.
+    // fn trigger_interrupt(&mut self) -> Result<(), T::E> {
+    //     // Triggers an event which is sent to the driver.
+    //     self.interrupt_evt.trigger()
+    // }
+
+    fn get_rtc_value(&self) -> u32 {
+        // Add the counter offset to the seconds elapsed since reset.
+        (self.counter.elapsed().as_secs() as u32) + self.lr
+    }
+
+    /// Handles a write request from the driver at `offset` offset from the
+    /// base register address.
+    ///
+    /// # Arguments
+    /// * `offset` - The offset from the base register specifying
+    ///              the register to be written.
+    /// * `data` - The little endian, 4 byte array to write to the register
+    ///
+    /// # Example
+    ///
+    /// You can see an example of how to use this function in the
+    /// [`Example` section from `RTC`](struct.RTC.html#example).
+    pub fn write(&mut self, offset: u64, data: &[u8; 4]) -> Result<(), Error<T::E>> {
```

Comment:
Can we define the offset as u32? It seems like it was define as an u64 because that's the size of the read on the Bus. But when implementing the bus, we should be able to check the offset, and only call into the `RTC::write` if it fits u32.

Reply:
Done.